### DISCUSSION TITLE
how to configure java adapter to load java-debug?

### DISCUSSION THREAD

**Discussion by joelpeapen** [mfussenegger/nvim-dap#1302] [Q&A]
The wiki says we can set up the java debugger independently of nvim-jdtls.

My LSP is already setup with lsp-config.
How exactly do i configure the java adapter to get the LSP to trigger the command `vscode.java.startDebugSession`?

Calling `dap.continue()` gives this error:
> Adapter used with attach must have a port property
> jdtls: -32601: No delegateCommandHandler for vscode.java.startDebugSession

**Comment by bayramtturgutt** [comment]
Hi @joelpeapen, pls follow below, hope works for you...

To integrate the Java debugger with your existing nvim-jdtls setup, you'll need to ensure proper configuration for both nvim-dap and the Java Debug Adapter. Here’s a structured approach to resolve the issues you’re encountering:

 ###    **Ensure Dependencies Are Installed:** 
    Verify that you have nvim-dap and nvim-dap-ui installed. These plugins are crucial for handling debugging sessions in Neovim.

 ###    **Configure the Java Debug Adapter:**
    Set up nvim-dap with the Java Debug Adapter. You need to specify the adapter's type, host, and port. Here’s a sample configuration:
    
```
    local dap = require('dap')

dap.adapters.java = {
  type = 'server',
  host = '127.0.0.1',
  port = 5005,
}

dap.configurations.java = {
  {
    type = 'java',
    name = 'Debug (Attach)',
    request = 'attach',
    hostName = '127.0.0.1',
    port = 5005,
  },
}

```
This setup tells nvim-dap where to find the Java Debug Adapter and how to connect to it.

### **Configure nvim-jdtls:**
Ensure nvim-jdtls is configured correctly to work with nvim-dap. Verify that your jdtls configuration aligns with your debugging setup:

```
local jdtls = require('jdtls')

jdtls.setup({
  init_options = {
    bundles = {},
  },
})
```
### **Launch Your Java Application for Debugging:**
Make sure your Java application is started with debugging enabled. Use the following command:
`
java -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005 -jar your-app.jar
`
Adjust the port if needed to match your nvim-dap configuration.


### **Trigger the Debug Session:**
From Neovim, you can start the debugging session using:
`require('dap').continue()`

Ensure the Java application is correctly listening on the port specified in your configuration.

### **Troubleshoot Common Errors:**

    Adapter used with attach must have a port property: Double-check that your dap.adapters.java configuration includes the port property.
    No delegateCommandHandler for vscode.java.startDebugSession: Confirm that the Java Debug Adapter is running and that nvim-dap is correctly configured to communicate with it.

**Reply by mfussenegger** [reply]
The server definition here won't work because nvim-dap can't connect directly to a jdwp agent.
See https://github.com/mfussenegger/nvim-dap/issues/451#issuecomment-2151512455 for details.

**Comment by mfussenegger** [comment]
> How exactly do i configure the java adapter to get the LSP to trigger the command vscode.java.startDebugSession?

> Calling dap.continue() gives this error:
> 
> Adapter used with attach must have a port property
> jdtls: -32601: No delegateCommandHandler for vscode.java.startDebugSession


The error indicates that the `java-debug` bundle is missing.
See https://github.com/mfussenegger/nvim-jdtls?tab=readme-ov-file#java-debug-installation for instructions on how to add it. (It doesn't require nvim-jdtls, you can set the `init_options`/`bundles` via lspconfig too)

**Comment by joelpeapen** [comment]
Turns out the problem with the bundles was i concatenated `vim.fn.stdpath("data") .. "/nvim/path/to/jar"` which returns `"~/.local/share/nvimnvim..."`

my adapter looks like this:
```lua
dap.adapters.java = function(callback)
    vim.lsp.buf.execute_command({
        command = "vscode.java.startDebugSession"
    }) -- not sure how to get the port from this command
    callback({
        type = "server";
        host = "127.0.0.1";
        port = 5005;
    })
end
```
which returns `Couldn't connect to 127.0.0.1:5005: ECONNREFUSED`

should i be using `vim.lsp.rpc.request()`?
